File: apps/agents/colorforge_agents/contracts/success_score.py
"""SuccessScore contract — output of Performance Monitor."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class SuccessScore(BaseModel):
    """Per-book success metric at a given time window."""

    book_id: str
    window_days: Literal[7, 14, 30]
    units_sold: int = Field(ge=0)
    royalty_total: float = Field(ge=0)
    kenp_read: int = Field(ge=0)
    refund_count: int = Field(ge=0)
    computed_score: float = Field(ge=0, le=100)
    classification: Literal["winner", "flat", "loser"]
    percentile_within_account: float = Field(ge=0, le=100)
    percentile_within_niche: float = Field(ge=0, le=100)


SUCCESS_SCORE_EXAMPLE = SuccessScore(
    book_id="770e8400-e29b-41d4-a716-446655440002",
    window_days=30,
    units_sold=47,
    royalty_total=84.32,
    kenp_read=1250,
    refund_count=1,
    computed_score=78,
    classification="winner",
    percentile_within_account=92,
    percentile_within_niche=88,
)

File: packages/shared/src/contracts/success-score.ts
import { z } from "zod";

export const SuccessScoreSchema = z.object({
  bookId: z.string().uuid(),
  windowDays: z.union([z.literal(7), z.literal(14), z.literal(30)]),
  unitsSold: z.number().int().nonnegative(),
  royaltyTotal: z.number().nonnegative(),
  kenpRead: z.number().int().nonnegative(),
  refundCount: z.number().int().nonnegative(),
  computedScore: z.number().min(0).max(100),
  classification: z.enum(["winner", "flat", "loser"]),
  percentileWithinAccount: z.number().min(0).max(100),
  percentileWithinNiche: z.number().min(0).max(100),
});

export type SuccessScore = z.infer<typeof SuccessScoreSchema>;

export const successScoreExample: SuccessScore = {
  bookId: "770e8400-e29b-41d4-a716-446655440002",
  windowDays: 30,
  unitsSold: 47,
  royaltyTotal: 84.32,
  kenpRead: 1250,
  refundCount: 1,
  computedScore: 78,
  classification: "winner",
  percentileWithinAccount: 92,
  percentileWithinNiche: 88,
};
